#### Loading the sql extension into the notebook

In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:password@localhost:3306/md_water_services

'Connected: root@md_water_services'

#### Show all tables in the database 

In [3]:
%sql SHOW TABLES FROM md_water_services

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('data_dictionary',),
 ('employee',),
 ('employee_copy',),
 ('global_water_access',),
 ('location',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',),
 ('well_pollution_copy',)]

Testing to make sure that the connection was done properly and its loading fully

In [6]:
%%sql

SELECT 
    *
FROM 
    visits
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

## 1. Looking into employees

#### Creating a new email for each employee 
The email addresses of each employee were initially not been added to the table. To be able to send employees reports and figures, these will be added and updated.

In [8]:
%%sql

SELECT 
    CONCAT(
    LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov') AS new_email /*The REPLACE function replaces the space with a full stop and everything is 
                                                                            put in lower case and then concatenated to make a single column*/
FROM 
    employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('amara.jengo@ndogowater.gov',),
 ('bello.azibo@ndogowater.gov',),
 ('bakari.iniko@ndogowater.gov',),
 ('malachi.mavuso@ndogowater.gov',),
 ('cheche.buhle@ndogowater.gov',),
 ('zuriel.matembo@ndogowater.gov',),
 ('deka.osumare@ndogowater.gov',),
 ('lalitha.kaburi@ndogowater.gov',),
 ('enitan.zuri@ndogowater.gov',),
 ('farai.nia@ndogowater.gov',)]

#### Updating the employee table with the result from the new emails just created 

In [10]:
%%sql

UPDATE employee
SET email = CONCAT(
    LOWER(REPLACE(employee_name, ' ', '.')), '@ndogowater.gov');

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

#### Testing to make sure the table was successfully updated

In [11]:
%%sql

SELECT * FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[(0, 'Amara Jengo', '+99637993287 ', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786 ', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041 ', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900 ', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640 ', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (5, 'Zuriel Matembo', '+99034075111 ', 'zuriel.matembo@ndogowater.gov', '26 Bahari Ya Faraja Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (6, 'Deka Osumare', '+99379364631 ', 'deka.osumare@ndogowater.gov', '104 Kenyatta Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (7, 'Lalitha Kaburi', '+99681623240 ', 'lalitha.kaburi@ndogowater.gov', '145 Sungura Amanpour Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (8, 'Enitan Zuri', '+99248509202 ', 'enitan.zuri@ndogowater.gov', '117 Kampala Road', 'Hawassa', 'Zanzibar', 'Field Surveyor'),
 (10, 'Farai Nia', '+99570082739 ', 'farai.nia@ndogowater.gov', '33 Angélique Kidjo Avenue', 'Amanzi', 'Dahabu', 'Field Surveyor')]

#### Removing excess characters at the end of the phone_numbers

In [20]:
%%sql

SELECT
phone_number,
RTRIM(phone_number) as new_phone_number,
LENGTH(RTRIM(phone_number))
FROM
employee;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


[('+99637993287 ', '+99637993287', 12),
 ('+99643864786 ', '+99643864786', 12),
 ('+99222599041 ', '+99222599041', 12),
 ('+99945849900 ', '+99945849900', 12),
 ('+99381679640 ', '+99381679640', 12),
 ('+99034075111 ', '+99034075111', 12),
 ('+99379364631 ', '+99379364631', 12),
 ('+99681623240 ', '+99681623240', 12),
 ('+99248509202 ', '+99248509202', 12),
 ('+99570082739 ', '+99570082739', 12),
 ('+99438283153 ', '+99438283153', 12),
 ('+99917230124 ', '+99917230124', 12),
 ('+99313603182 ', '+99313603182', 12),
 ('+99820596465 ', '+99820596465', 12),
 ('+99176320477 ', '+99176320477', 12),
 ('+99611183730 ', '+99611183730', 12),
 ('+99690308755 ', '+99690308755', 12),
 ('+99522694997 ', '+99522694997', 12),
 ('+99797494944 ', '+99797494944', 12),
 ('+99822478933 ', '+99822478933', 12),
 ('+99533195846 ', '+99533195846', 12),
 ('+99046972648 ', '+99046972648', 12),
 ('+99477692836 ', '+99477692836', 12),
 ('+99895654233 ', '+99895654233', 12),
 ('+99613194771 ', '+99613194771', 12),
 ('+99489752164 ', '+99489752164', 12),
 ('+99269319001 ', '+99269319001', 12),
 ('+99239172440 ', '+99239172440', 12),
 ('+99239164858 ', '+99239164858', 12),
 ('+99578401238 ', '+99578401238', 12),
 ('+99345162378 ', '+99345162378', 12),
 ('+99239271483 ', '+99239271483', 12),
 ('+99738251476 ', '+99738251476', 12),
 ('+99471523698 ', '+99471523698', 12),
 ('+99381254763 ', '+99381254763', 12),
 ('+99142537684 ', '+99142537684', 12),
 ('+99924513876 ', '+99924513876', 12),
 ('+99568247135 ', '+99568247135', 12),
 ('+99735681472 ', '+99735681472', 12),
 ('+99317854629 ', '+99317854629', 12),
 ('+99874512369 ', '+99874512369', 12),
 ('+99127436598 ', '+99127436598', 12),
 ('+99583124796 ', '+99583124796', 12),
 ('+99273841596 ', '+99273841596', 12),
 ('+99968217435 ', '+99968217435', 12),
 ('+99682314579 ', '+99682314579', 12),
 ('+99721534698 ', '+99721534698', 12),
 ('+99578123469 ', '+99578123469', 12),
 ('+99375418692 ', '+99375418692', 12),
 ('+99178325649 ', '+99178325649', 12),
 ('+99234156789 ', '+99234156789', 12),
 ('+99712584936 ', '+99712584936', 12),
 ('+99634157892 ', '+99634157892', 12),
 ('+99231569847 ', '+99231569847', 12),
 ('+99123456978 ', '+99123456978', 12),
 ('+99412356789 ', '+99412356789', 12)]

#### Creating a duplicate employee table to be able to make sure the results are good before we update the main employee table 

In [21]:
%%sql 

CREATE TABLE
    employee_copy
AS(
        SELECT *
        FROM employee
);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

Before the Update 

In [54]:
%%sql 

SELECT * FROM employee_copy LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

#### Making the update on the employee_copy table

In [23]:
%%sql 

UPDATE employee_copy
SET phone_number = RTRIM(phone_number);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

#### After the update to make sure it ran successfully 

In [55]:
%%sql 

SELECT * FROM employee_copy LIMIT 5

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

In [57]:
%%sql

SELECT 
    phone_number,
    LENGTH(phone_number)
FROM
    employee_copy
LIMIT 5;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('+99637993287', 12),
 ('+99643864786', 12),
 ('+99222599041', 12),
 ('+99945849900', 12),
 ('+99381679640', 12)]

#### Finding the number of employees who live in particular areas

In [59]:
%%sql 

SELECT 
    DISTINCT(town_name),
    COUNT(employee_name) OVER (
        PARTITION BY town_name) AS num_employees
FROM
    employee
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


[('Dahabu', 6),
 ('Harare', 5),
 ('Ilanga', 3),
 ('Kintampo', 1),
 ('Lusaka', 4),
 ('Rural', 29),
 ('Serowe', 3),
 ('Yaounde', 1),
 ('Zanzibar', 4)]

In [13]:
%%sql 

SELECT 
    *
FROM 
    employee
WHERE 
    town_name = 'Harare'
    AND province_name = 'Kilimani';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[(58, 'Faridah Khalida', '+99142537684 ', 'faridah.khalida@ndogowater.gov', '176 Port Said Street', 'Kilimani', 'Harare', 'Pioneer'),
 (87, 'Pili Xola', '+99682314579 ', 'pili.xola@ndogowater.gov', '31 Samora Machel Road', 'Kilimani', 'Harare', 'Civil Engineer')]

We can see that 6 people live in Dahabu

#### Finding the number of visits made by each employee and sorting to get the top 3 with the most number of visits

In [80]:
%%sql 

SELECT *
FROM visits
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46)]

In [33]:
%%sql 

SELECT
    DISTINCT(assigned_employee_id),
    COUNT(visit_count) OVER (PARTITION BY assigned_employee_id) as num_of_visits
FROM
    visits
 ORDER BY num_of_visits DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(1, 3708), (30, 3676), (34, 3539)]

The query below does the same thing as the query above. 

In [21]:
%%sql 

SELECT
    DISTINCT(assigned_employee_id),
    COUNT(visit_count) as num_of_visits
FROM
    visits
GROUP BY assigned_employee_id
ORDER BY num_of_visits DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(1, 3708), (30, 3676), (34, 3539)]

### Finding the employees who had the lowest visits and their names 

In [81]:
%%sql 

SELECT
    DISTINCT(assigned_employee_id),
    COUNT(visit_count) as num_of_visits
FROM
    visits
GROUP BY assigned_employee_id
ORDER BY num_of_visits ASC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(20, 15), (22, 143), (44, 218)]

In [8]:
%%sql 

SELECT * FROM visits
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46)]

In [7]:
%%sql 

SELECT * FROM employee_copy
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor')]

In [83]:
%%sql 

SELECT 
    assigned_employee_id,
    employee_name
FROM
    employee_copy
WHERE 
    assigned_employee_id IN (20, 22, 44);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(20, 'Kunto Asha'), (22, 'Lesedi Kofi'), (44, 'Wambui Jabali')]

In [9]:
%%sql
 
SELECT 
    v.assigned_employee_id,
    COUNT(visit_count) as num_of_visits,
    emp.employee_name
FROM 
    visits as v 
INNER JOIN 
    employee as emp
    ON v.assigned_employee_id = emp.assigned_employee_id
GROUP BY 
    assigned_employee_id
ORDER BY 
    num_of_visits ASC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
29 rows affected.


[(20, 15, 'Kunto Asha'),
 (22, 143, 'Lesedi Kofi'),
 (44, 218, 'Wambui Jabali'),
 (26, 719, 'Nia Furaha'),
 (16, 802, 'Isoke Amani'),
 (0, 1099, 'Amara Jengo'),
 (4, 1136, 'Cheche Buhle'),
 (6, 1182, 'Deka Osumare'),
 (7, 1259, 'Lalitha Kaburi'),
 (14, 1441, 'Harith Nyota'),
 (46, 1632, 'Xola Uzuri'),
 (18, 1749, 'Jengo Tumaini'),
 (32, 1852, 'Quibilah Akua'),
 (24, 2015, 'Makena Thabo'),
 (2, 2033, 'Bakari Iniko'),
 (38, 2121, 'Thandiwe Kito'),
 (40, 2344, 'Usafi Ayo'),
 (42, 2496, 'Vuyisile Kwame'),
 (12, 2561, 'Gamba Shani'),
 (28, 2762, 'Ona Sefu'),
 (48, 2933, 'Yewande Ebele'),
 (36, 3249, 'Sanaa Tendaji'),
 (5, 3284, 'Zuriel Matembo'),
 (8, 3351, 'Enitan Zuri'),
 (10, 3407, 'Farai Nia'),
 (3, 3420, 'Malachi Mavuso'),
 (34, 3539, 'Rudo Imani'),
 (30, 3676, 'Pili Zola'),
 (1, 3708, 'Bello Azibo')]

## 2. Having a look at the 'location' table

In [11]:
%%sql 

SELECT 
    province_name,
    town_name,
    location_type
FROM location 
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Akatsi', 'Harare', 'Urban'), ('Akatsi', 'Harare', 'Urban')]

#### Find out the number of records gotten per town 

In [13]:
%%sql 

SELECT 
    town_name,
    COUNT(town_name) AS records_per_town
FROM
    location
GROUP BY town_name
ORDER BY records_per_town DESC
LIMIT 6;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
6 rows affected.


[('Rural', 23740),
 ('Harare', 1650),
 ('Amina', 1090),
 ('Lusaka', 1070),
 ('Mrembo', 990),
 ('Asmara', 930)]

#### Discovering the number of records that were gotten at the province level

In [15]:
%%sql 

SELECT 
    province_name,
    COUNT(province_name) AS records_per_province
FROM
    location
GROUP BY province_name
ORDER BY records_per_province DESC
LIMIT 6;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Kilimani', 9510),
 ('Akatsi', 8940),
 ('Sokoto', 8220),
 ('Amanzi', 6950),
 ('Hawassa', 6030)]

#### Getting the number of records in each town, within each province and ordering it based on the highest number per province 

In [22]:
%%sql 

SELECT 
    province_name,
    town_name,
    COUNT(town_name) AS records_per_town
FROM
    location
GROUP BY 
    province_name,
    town_name
ORDER BY 
    province_name,
    records_per_town DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Akatsi', 'Lusaka', 1070),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Amanzi', 'Rural', 3100),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Amanzi', 'Amina', 670),
 ('Amanzi', 'Pwani', 520),
 ('Amanzi', 'Abidjan', 400)]

#### Looking at the number of records per location_type 

In [19]:
%%sql 

SELECT 
    location_type,
    COUNT(location_type) AS num_of_sources
FROM
    location
GROUP BY location_type
LIMIT 6;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Urban', 15910), ('Rural', 23740)]

## 3. Diving into the 'water_source' table

In [25]:
%%sql 

SELECT *
FROM water_source
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486)]

#### Finding the total number of people that were surveyed from Maji Ndogo

In [40]:
%%sql 

SELECT 
    SUM(number_of_people_served) as total_num_served
FROM 
    water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('27628140'),)]

#### Finding the type of water sources that are available and how many they are

In [31]:
%%sql 

SELECT 
    type_of_water_source,
    COUNT(type_of_water_source) as number_of_sources
FROM 
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    number_of_sources DESC;
    

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('well', 17383),
 ('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('shared_tap', 5767),
 ('river', 3379)]

In [39]:
%%sql 

SELECT 
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) as AVG_number_of_people
FROM 
    water_source
GROUP BY 
    type_of_water_source
ORDER BY 
    type_of_water_source
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('river', Decimal('699')),
 ('shared_tap', Decimal('2071')),
 ('tap_in_home', Decimal('644')),
 ('tap_in_home_broken', Decimal('649')),
 ('well', Decimal('279'))]

We can see that an average of 279 people share a well.


##### For tap_in_home and tap_in_home_broken.

Records have been joined together and there are about 6 people living in a given household. Each household has a single tap in it.

This means that 1 tap_in_home actually represents 644 ÷ 6 = ± 100 taps.

#### Getting  the number of people served by each water source

In [42]:
%%sql 

SELECT 
    type_of_water_source,
    SUM(number_of_people_served) as population_served
FROM 
    water_source
GROUP BY
    type_of_water_source
ORDER BY 
    population_served DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272')),
 ('well', Decimal('4841724')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('river', Decimal('2362544'))]

Calculating the total amount of people who use a water_source like a tap.

In [53]:
%%sql 

SELECT 
    type_of_water_source,
    SUM(number_of_people_served) as population_served
FROM 
    water_source
GROUP BY
    type_of_water_source
HAVING
    type_of_water_source LIKE '%tap%'
ORDER BY 
    population_served DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[('shared_tap', Decimal('11945272')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720'))]

#### Converting the population per source into percentages so as to be able to interpret the numbers accordingly

In [48]:
%%sql 

SELECT 
    type_of_water_source,
    ROUND((SUM(number_of_people_served)/27628140) * 100, 0) as percentage_num_served
FROM 
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    percentage_num_served DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('43')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

## 4. Data driven solution begins

In [101]:
%%sql 

SELECT 
    type_of_water_source,
    SUM(number_of_people_served) as population_served,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) as rank_served
FROM 
    water_source
GROUP BY
    type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home', Decimal('4678880'), 3),
 ('tap_in_home_broken', Decimal('3799720'), 4),
 ('river', Decimal('2362544'), 5)]

In [107]:
%%sql

SELECT *
FROM water_source
LIMIT 2;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930)]

#### Comparing the different water sources and giving them priority numbers to know which should be handled first based on the number of people that utilize them.

In [128]:
%%sql 

SELECT 
    source_id,
    type_of_water_source,
    number_of_people_served,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) as priority_rank
FROM
    water_source
GROUP BY 
    source_id
ORDER BY 
    source_id
LIMIT 105;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
105 rows affected.


[('AkHa00000224', 'tap_in_home', 956, 5564),
 ('AkHa00001224', 'tap_in_home_broken', 930, 6225),
 ('AkHa00002224', 'tap_in_home_broken', 486, 17692),
 ('AkHa00003224', 'well', 364, 23234),
 ('AkHa00004224', 'tap_in_home_broken', 942, 5930),
 ('AkHa00005224', 'tap_in_home', 736, 11273),
 ('AkHa00006224', 'tap_in_home', 882, 7459),
 ('AkHa00007224', 'tap_in_home', 554, 15913),
 ('AkHa00008224', 'well', 398, 20047),
 ('AkHa00009224', 'well', 346, 24880),
 ('AkHa00010224', 'well', 236, 33837),
 ('AkHa00011224', 'well', 168, 38948),
 ('AkHa00012224', 'well', 192, 37192),
 ('AkHa00013224', 'shared_tap', 2212, 2624),
 ('AkHa00014224', 'tap_in_home_broken', 574, 15414),
 ('AkHa00015224', 'tap_in_home', 862, 7941),
 ('AkHa00016224', 'well', 278, 30758),
 ('AkHa00017224', 'tap_in_home_broken', 414, 19618),
 ('AkHa00018224', 'well', 390, 20803),
 ('AkHa00019224', 'tap_in_home_broken', 320, 27249),
 ('AkHa00020224', 'well', 234, 34002),
 ('AkHa00021224', 'tap_in_home_broken', 330, 26318),
 ('AkHa00022224', 'well', 304, 28682),
 ('AkHa00023224', 'tap_in_home_broken', 954, 5622),
 ('AkHa00024224', 'well', 292, 29730),
 ('AkHa00025224', 'tap_in_home', 354, 24146),
 ('AkHa00026224', 'well', 388, 21005),
 ('AkHa00027224', 'well', 224, 34764),
 ('AkHa00028224', 'well', 306, 28509),
 ('AkHa00029224', 'tap_in_home_broken', 816, 9164),
 ('AkHa00030224', 'well', 366, 23022),
 ('AkHa00031224', 'well', 354, 24146),
 ('AkHa00032224', 'well', 382, 21542),
 ('AkHa00033224', 'well', 384, 21375),
 ('AkHa00034224', 'well', 210, 35861),
 ('AkHa00035224', 'tap_in_home_broken', 920, 6479),
 ('AkHa00036224', 'shared_tap', 3488, 691),
 ('AkHa00037224', 'well', 348, 24695),
 ('AkHa00038224', 'well', 308, 28328),
 ('AkHa00039224', 'well', 252, 32693),
 ('AkHa00040224', 'tap_in_home_broken', 338, 25608),
 ('AkHa00041224', 'tap_in_home', 672, 12880),
 ('AkHa00042224', 'tap_in_home', 766, 10504),
 ('AkHa00043224', 'tap_in_home_broken', 364, 23234),
 ('AkHa00044224', 'well', 304, 28682),
 ('AkHa00045224', 'well', 370, 22674),
 ('AkHa00046224', 'well', 186, 37661),
 ('AkHa00047224', 'well', 324, 26883),
 ('AkHa00048224', 'well', 174, 38507),
 ('AkHa00049224', 'tap_in_home_broken', 892, 7194),
 ('AkHa00050224', 'tap_in_home_broken', 716, 11793),
 ('AkHa00051224', 'tap_in_home', 900, 6970),
 ('AkHa00052224', 'tap_in_home', 374, 22284),
 ('AkHa00053224', 'tap_in_home_broken', 748, 10960),
 ('AkHa00054224', 'well', 378, 21919),
 ('AkHa00055224', 'tap_in_home', 562, 15724),
 ('AkHa00056224', 'well', 328, 26493),
 ('AkHa00057224', 'tap_in_home_broken', 876, 7602),
 ('AkHa00058224', 'well', 290, 29896),
 ('AkHa00059224', 'well', 316, 27619),
 ('AkHa00060224', 'tap_in_home_broken', 376, 22103),
 ('AkHa00061224', 'tap_in_home', 908, 6758),
 ('AkHa00062224', 'tap_in_home_broken', 576, 15369),
 ('AkHa00063224', 'well', 306, 28509),
 ('AkHa00064224', 'tap_in_home_broken', 538, 16328),
 ('AkHa00065224', 'tap_in_home_broken', 874, 7650),
 ('AkHa00066224', 'tap_in_home', 722, 11643),
 ('AkHa00067224', 'well', 302, 28877),
 ('AkHa00068224', 'well', 264, 31787),
 ('AkHa00069224', 'well', 306, 28509),
 ('AkHa00070224', 'well', 390, 20803),
 ('AkHa00071224', 'well', 212, 35687),
 ('AkHa00072224', 'tap_in_home_broken', 816, 9164),
 ('AkHa00073224', 'well', 354, 24146),
 ('AkHa00074224', 'tap_in_home_broken', 322, 27074),
 ('AkHa00075224', 'tap_in_home_broken', 358, 23786),
 ('AkHa00076224', 'well', 298, 29311),
 ('AkHa00077224', 'well', 262, 31922),
 ('AkHa00078224', 'tap_in_home_broken', 612, 14470),
 ('AkHa00079224', 'tap_in_home', 332, 26147),
 ('AkHa00080224', 'tap_in_home', 880, 7507),
 ('AkHa00081224', 'well', 318, 27446),
 ('AkHa00082224', 'tap_in_home_broken', 322, 27074),
 ('AkHa00083224', 'tap_in_home_broken', 744, 11058),
 ('AkHa00084224', 'shared_tap', 322, 27074),
 ('AkHa00085224', 'well', 346, 24880),
 ('AkHa00086224', 'well', 206, 36164),
 ('AkHa00087224', 'well', 288, 30025),
 ('AkHa00088224', 'well', 354, 24146),
 ('AkHa00089224', 'well', 268, 31498),
 ('AkHa00090224', '